### ISSUE TITLE
[MdAutocomplete] options with objects.

### ISSUE THREAD

**Issue by VdustR** [vuematerial/vue-material#1243]
As #1215, I fixed by #1218 to make `searchTerm` sync with `v-model`, but that also make it crash while the options are objects. It's could reproduce in documentation example **Custom Template** - **With highlight text** with 6f51eb91338a0fa845a99b2a65f1c76f8d549196 (and it's a fatal error now in beta 6). If I want to make #1215 support object, I should know what to show as `searchTerm`. I think it's not easy because `MdAutocomplete` get the plain text via [`Node.textContent`](https://github.com/vuematerial/vue-material/blob/cb795934daab18df84b22e11c3be8675209bf6a1/src/components/MdAutocomplete/MdAutocomplete.vue#L206).

Another issue is that `MdIcon` cannot be used in `md-autocomplete-item` slots or the `searchTerm` will show the text in `MdIcon`.

As #1241, I think users prefer that the result is the thing shown on the `searchTerm` not the thing they select via autocomplete. It's good if I could still get the object while I type the `searchTerm` and match a statement.

Any idea?

**Comment by mordonez-me** [comment]
Is there a workaround for this?

**Comment by borekwa** [comment]
Bump.  I had to resort to using another autocomplete lib because I couldn't workaround this.

**Comment by Xilesun** [comment]
I try this:

```
<md-autocomplete v-model="searchvalue" :md-options="lists" :disabled="sending" @md-selected="select" @md-opened="select">
   <label>List</label>
   <template slot="md-autocomplete-item" slot-scope="{ item }">{{ item.text }}</template>
</md-autocomplete>
```

where `searchvalue` is the `searchTearm` and the `item.text` is the text displayed and match the `searchTerm`

then the `select` function:

```
selectRepo(v) {
    if (!v) {
      v = {value: '', text: ''}
    }
    this.searchvalue = v.text
    this.realvalue = v.value
}
```

where `realvalue` and `list.value` is the result you selected.

**Comment by rkakrik** [comment]
Same problem

**Comment by fizxmike** [comment]
@rkakrik you solution did not work for me. The @md-selected="onSelect" will grab the right value, but md-autocomplet component still throws errors if the objects in you list are not "stringy" ... so you have to implement .toLowerCase and .toString on each (or maybe you are already working with "proper" objects - I wasn't). For example (hacky/ugly, but works, and maintained re-usability for me to just put object mapping in md-options):
```vue
<md-autocomplete
  v-model="SelectDialog.SelectedItem"
  :md-options="SelectDialog.SelectList.map(x=>({
    'Id':x.Id,
    'Name':x.Name,
    'toLowerCase':()=>x.Name.toLowerCase(),
    'toString':()=>x.Name
  }))"
>
  <label>Select</label>

  <template slot="md-autocomplete-item" slot-scope="{ item, term }">
    <md-highlight-text :md-term="term">{{ item }}</md-highlight-text>
  </template>

  <template slot="md-autocomplete-empty" slot-scope="{ term }">
    "{{ term }}" not found!
  </template>

  <div class="md-helper-text" v-if="SelectDialog.SelectedItem">
    <strong>Item:</strong>{{SelectDialog.SelectedItem.Name}} ({{SelectDialog.SelectedItem.Id}})
  </div>
</md-autocomplete>
```

**Comment by fizxmike** [comment]
... the silly thing is that the native <select> in html already had this concept of "value" and "label" separate...

I'd suggest leveraging that somehow with a slot (somehow). The v-model should be the selected OBJECT, and the text should come from selected object's "label" (defined by slot)? Not sure if it's possible.

Alternatively, you can add a prop for one to pass a "labeler" function (the default being .toString). Labeler function is evaluated to show list items and set text in text box. V-model would still hold ref to selected OBJECT, not the label.

**Comment by Samuell1** [comment]
@fizxmike https://github.com/vuematerial/vue-material/pull/1338

**Comment by DavidHtx** [comment]
@fizxmike your solution partially worked for me, instead of the item name showing it was showing the entire object. 

My workaround was as follows:

```html
  <md-autocomplete 
                   v-model="selectedLabel"
                   :md-options="getSanitizedLabels">
    <label>Label</label>

    <template slot="md-autocomplete-item" slot-scope="{ item, term }">
      <md-highlight-text :md-term="term">{{ item.Name }}</md-highlight-text>
    </template>

  </md-autocomplete>

```
```javascript
computed: {
    getSanitizedLabels() {
      return this.labels.map(label => ({
        'Id': label.id,
        'Name': label.name,
        'toLowerCase': () => label.name.toLowerCase(),
        'toString': () => label.name
      }));
    }
  }
```

Although this works for me, it gives an error message when selecting an item:
`[Vue warn]: Invalid prop: type check failed for prop "mdTerm". Expected String, got Object.`

**Comment by fizxmike** [comment]
Yea... this, and other issues re-inventing (bad) wheels, have "drove" me to Angular (TypeScript) with [Angular Material components](https://material.angular.io/components/select/overview). Google's "wheels" tend to work just fine ([most of the time](https://github.com/angular/material2/issues/8420#issuecomment-411078672)).

**Comment by vzapo** [comment]
Hi all,

I'm facing the same issue, solution provided by @DavidHtx and @fizxmike works but both partially, as in it works passing object and remaping using the promise, but it will still throw : [Vue warn]: Invalid prop: type check failed for prop "mdTerm". Expected String, got Object.
if you want to select something the second time from the list. Still works thou but chrome console throws it.

Any idea when this will be fixed or if it will be fixed?

**Comment by samuelterra22** [comment]
@vzapo try add  `term.Name` in  `md-term`
```
 <template slot="md-autocomplete-item" slot-scope="{ item, term }">
      <md-highlight-text :md-term="term.Name">{{ item.Name }}</md-highlight-text>
    </template>
```

**Comment by vzapo** [comment]
> thanks.
> 
> @vzapo try add `term.Name` in `md-term`
> 
> ```
>  <template slot="md-autocomplete-item" slot-scope="{ item, term }">
>       <md-highlight-text :md-term="term.Name">{{ item.Name }}</md-highlight-text>
>     </template>
> ```

**Comment by gaspo53** [comment]
Same issue here, I've tried everything, and still getting `this.searchTerm.toLowerCase` is not a function.

![image](https://user-images.githubusercontent.com/2250227/56436568-b14fab80-62b2-11e9-9f9a-74bc56d9593f.png)


Is vue-material still supported? Or should we switch to another component?

Thanks

**Comment by koAndy** [comment]
Any news with this issue? Including autocomplete on documentation website is not working properly objects. (See Custom template with highlight text, where you cannot select an option and then everything breaks)

**Comment by manu144x** [comment]
I resorted to not using v-model, and using :value and @input and handling it myself.

However now the issue is that the value prop is not watched, so when I'm updating it manually (resetting it to empty basically after I did my thing) it's not resetting also on the component.

So still unusable.

If this would work, at least we'd have that way of working around this bug

**Comment by uicestone** [comment]
If you read the source code of MdAutocomplete component you will find v-model is bind to "searchTerm". The author means to bind the search keyword to v-model of this input rather than the result string/object it self. So if we follow this idea, use xxxSearchTerm, onSelected trigger to update our actual data. Just like @rkakrik did, but you need to bind another searchTerm to v-model, not the form field it self.

For the clear form issue, I don't think it's a problem. Because whether the searchTerm is part of form or not is just your choice. In my use case, I just put the search term out of form.

```vue
<md-autocomplete
  v-model="customerSearchTerm"
  :md-options="customers"
  @md-changed="getCustomers"
  @md-opened="getCustomers"
  @md-selected="selectCustomer"
>
  <label>Customer</label>

  <template slot="md-autocomplete-item" slot-scope="{ item }">
    {{ item }}
  </template>
</md-autocomplete>

```
```js
  data() {
    return {
      booking: { id: "", customer: {} },
      customers: [],
      customerSearchTerm: ""
    };
  },
```
```js
    async getCustomers(q) {
      this.customers = (await User.get({ keyword: q })).body.map(c => c.name);
    },
    selectCustomer(c, b, a) {
      console.log(c, b, a);
      this.booking.customer = c;
    }

```

**Comment by uicestone** [comment]
Separate object field and search term, change search term onSelected, it works!

If you read the source code of MdAutocomplete component you will find v-model is bind to "searchTerm". The author means to bind the search keyword to v-model of this input rather than the result string/object it self. So if we follow this idea, use xxxSearchTerm, onSelected trigger to update our actual data. Just like @rkakrik did, but you need to bind another searchTerm to v-model, not the form field it self.

For the clear form issue, I don't think it's a problem. Because whether the searchTerm is part of form or not is just your choice. In my use case, I just put the search term out of form.

```vue
<md-autocomplete
  v-model="customerSearchTerm"
  :md-options="customers"
  @md-changed="getCustomers"
  @md-opened="getCustomers"
  @md-selected="selectCustomer"
>
  <label>Customer</label>

  <template slot="md-autocomplete-item" slot-scope="{ item }">
    {{ item }}
  </template>
</md-autocomplete>

```
```js
  data() {
    return {
      booking: { id: "", customer: {} },
      customers: [],
      customerSearchTerm: ""
    };
  },
```
```js
  methods: {
    async getCustomers(q) {
      this.customers = (await User.get({ keyword: q })).body;
    },
    selectCustomer(c) {
      this.booking.customer = c;
      this.customerSearchTerm = c.name;
    }
  },
  async mounted() {
    if (this.$route.params.id !== "add") {
      this.booking = (await Booking.get({ id: this.$route.params.id })).body;
      this.customerSearchTerm = this.booking.customer.name;
    }
  }

```

**Comment by Byronsargeant** [comment]
> @fizxmike your solution partially worked for me, instead of the item name showing it was showing the entire object.
> 
> My workaround was as follows:
> 
> ```
>   <md-autocomplete 
>                    v-model="selectedLabel"
>                    :md-options="getSanitizedLabels">
>     <label>Label</label>
> 
>     <template slot="md-autocomplete-item" slot-scope="{ item, term }">
>       <md-highlight-text :md-term="term">{{ item.Name }}</md-highlight-text>
>     </template>
> 
>   </md-autocomplete>
> ```
> 
> ```js
> computed: {
>     getSanitizedLabels() {
>       return this.labels.map(label => ({
>         'Id': label.id,
>         'Name': label.name,
>         'toLowerCase': () => label.name.toLowerCase(),
>         'toString': () => label.name
>       }));
>     }
>   }
> ```
> 
> Although this works for me, it gives an error message when selecting an item:
> `[Vue warn]: Invalid prop: type check failed for prop "mdTerm". Expected String, got Object.`

This seems to work for me without any error messages!

**Comment by anurat** [comment]
@uicestone  
Thank you for the solution. Now it doesn't throw errors anymore.  
However I am able to search only once.  

After I search and select an option from the option list for the first time, I remove the search term and try again, this time option list do not show up.  
I also looked at the options data in vue devtools and the options are retrieved correctly as usual, the UI just doesn't show up.  

Any suggestion would be appreciated.  
Thanks

**Comment by matias-miranda** [comment]
@anurat probably a bit late, but I was able to make it work adding the following

`[
    {
      "uuid": "12345-61c5-4bba-a3bb-4614c10283e3",
      "person_dni": 32109344,
      toLowerCase() {
         return this.uuid
      },
      toString() {
         return this.person_dni
      }
    },
    {
      "uuid": "935744-fd96-438c-85ae-d1391f9c54f2",
      "person_dni": 1234567,
      toLowerCase() {
         return this.uuid
      },
      toString() {
         return this.person_dni
      }
    }
  ]`

If I return the same value for toLowerCase() and toString() the dropdown is shown only once... I didn't investigate the root cause, but this workaround works fine

**Comment by nonghaiqn195** [comment]
![Screenshot from 2021-07-17 17-31-17](https://user-images.githubusercontent.com/52830944/126034196-19a730ba-49a3-42e6-8baa-edf02c3ae9f5.png)
Even the document does not solve the problem
